### ISSUE TITLE
Transformers mega issue

### ISSUE THREAD

**Issue by coreylowman** [coreylowman/dfdx#34]
Would like to add an small example of using a transformer architecture. This will likely involve new features such as batch mat mul and maybe some others.

**Comment by coreylowman** [comment]
https://towardsdatascience.com/how-to-code-the-transformer-in-pytorch-24db27c8f9ec is a good description of all the parts

**Comment by coreylowman** [comment]
Single head attention:

```rust
struct SingleHeadAttention<const M: usize, const K: usize, const V: usize> {
    w_q: Tensor2D<M, K, NoTape>,
    w_k: Tensor2D<M, K, NoTape>,
    w_v: Tensor2D<M, V, NoTape>,
}
```

where forward would take `Tensor2D<B, M>` and do matrix multiplication which each of the weights to extract q/k/v

**Comment by coreylowman** [comment]
Multi head attention:

```rust
struct MultiHeadAttention<const M: usize, const K: usize, const V: usize, const N: usize> {
    heads: [SingleHeadAttention<M, K, V>; N],
    w_out: Tensor2D<N * V, M, NoTape>,
}
```

Where forward would take `Tensor2D<B, M>`, duplicate the input N times and pass through each head's forward, then call concat (#43), and then matrix multiply against w_out

**Comment by coreylowman** [comment]
> For first pass MultiHeadAttention could just take `heads: SingleHeadAttention<M, NK, NV>` and `w_out: Tensor2D<NV, M>`, then this wouldn't have to depend on concat

Hmm but is N MxK matrices the same as 1 MxNK matrix for matmul?

**Comment by coreylowman** [comment]
Okay the key for multi head is that the softmax needs to be applied per head. So the above version where we do `SingleHeadAttention<M, N * K, N * V>` doesn't work because then softmax would be applied once across all the heads.

**Comment by coreylowman** [comment]
Additionally concat (#43) for multi head attention needs to be applied across the last dimension (e.g. NumHeads tensors of shape (B, V) need to be concatted to form 1 tensor of shape (B, NumHeads * V)

**Comment by coreylowman** [comment]
Another useful reference: https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py#L186

**Comment by jafioti** [comment]
@coreylowman Wouldn't dynamically sized dimensions be needed for transformers since it needs to take in variable length sequences?

**Comment by coreylowman** [comment]
@jafioti In practice it seems like most transformer libraries specify a max sequence length, and pad all sequences to that length during training & batch inference. Agree that true variable length sequences of any size would need dynamic dimension, but I think max sequence length is a good place to start for dfdx and we can go from there. I want to push const generics as far as they can go for DL, but if things aren't possible then will have to reevaluate at that time!

Another aside, but vision transformers should not have this problem since the number of image patches generated from an image of a given size is fixed. So those should be compatible as well!

**Comment by jafioti** [comment]
@coreylowman One thing I tried a while back was rather than using usize for tensor dim, I used an enum:
```rust
enum DimType {
    Static(usize),
    Dynamic,
}
```
The main downside is that you can't specify equality correctly yet, for instance you should be able to specify Static == Dynamic, Dynamic == Static, but Static(a) != Static(b) if a != b. Currently you can only derive PartialEq for any types used as const generics, so you can't implement custom equality logic.

Also it gets hella verbose.

**Comment by jafioti** [comment]
I made an initial implementation of SingleHeadAttention, but I'm not sure if it actually works lol. I'll need to test it out. The softmax should definitely have a dimension parameter.
```rust
#[derive(Debug, Clone)]
pub struct SingleHeadAttention<const M: usize, const K: usize, const V: usize> {
    w_q: Tensor2D<M, K, NoneTape>,
    w_k: Tensor2D<M, K, NoneTape>,
    w_v: Tensor2D<M, V, NoneTape>,
}

impl<const M: usize, const K: usize, const V: usize, const B: usize> Module<Tensor2D<B, M>> for SingleHeadAttention<M, K, V> {
    type Output = Tensor2D<B, V>;

    fn forward(&self, input: Tensor2D<B, M>) -> Self::Output {
        let queries = matmul(input.duplicate(), &self.w_q);
        let keys = matmul(input.duplicate(), &self.w_k);
        let values = matmul(input, &self.w_v);

        // Get weights
        let token_weights = matmul_transpose(queries, &keys) / (M as f32);

        // Softmax on last dimension
        let token_weights = softmax(token_weights);

        // Get new tokens
        matmul(token_weights, &values)
    }
}
```

**Comment by jafioti** [comment]
Actually scratch that, we don't even need a transformer block. Just use a tuple! Here's a full TransformerEncoder:
```rust
#[derive(Debug, Clone)]
pub struct TransformerEncoder<const M: usize, const I: usize, const K: usize, const L: usize> {
    blocks: Repeated<
        (
            SingleHeadAttention<M, K, M>,
            Linear<M, I>,
            ReLU,
            Linear<I, M>,
        ),
        L,
    >,
}

impl<const S: usize, const M: usize, const I: usize, const K: usize, const L: usize> Module<Tensor2D<S, M>>
    for TransformerEncoder<M, I, K, L>
{
    type Output = Tensor2D<S, M>;

    fn forward(&self, input: Tensor2D<S, M>) -> Self::Output {
        self.blocks.forward(input)
    }
}
```

**Comment by jafioti** [comment]
Basically the only things needed are ways to split a tensor on the last dimension and concat on the last dimension.

**Comment by jafioti** [comment]
I was able to take the SingleHeadAttention and implement both normal self attention as well as encoder-decoder style attention:
```rust
/// Normal self attention (where same tensors are used for keys, queries and values)
impl<const M: usize, const K: usize, const V: usize, const B: usize> Module<Tensor2D<B, M>>
    for SingleHeadAttention<M, M, K, V>
{
    type Output = Tensor2D<B, V>;

    fn forward(&self, input: Tensor2D<B, M>) -> Self::Output {
        let queries = matmul(input.duplicate(), &self.w_q);
        let keys = matmul(input.duplicate(), &self.w_k);
        let values = matmul(input, &self.w_v);

        // Get weights
        let token_weights = matmul_transpose(queries, &keys) / (M as f32);

        // Softmax on last dimension
        let token_weights = softmax(token_weights);

        // Get new tokens
        matmul(token_weights, &values)
    }
}

/// Encoder-Decoder style self attention where one set of tensors is used for values and keys, and another is used for queries
impl<const M: usize, const N: usize, const K: usize, const V: usize, const B: usize, const S: usize> Module<(Tensor2D<S, M>, Tensor2D<B, N>)>
    for SingleHeadAttention<M, N, K, V>
{
    type Output = Tensor2D<B, V>;

    fn forward(&self, (from_enc, input): (Tensor2D<S, M>, Tensor2D<B, N>)) -> Self::Output {
        let queries = matmul(input.duplicate(), &self.w_q);
        let keys = matmul(from_enc.duplicate(), &self.w_k);
        let values = matmul(from_enc, &self.w_v);

        // Get weights
        let token_weights = matmul_transpose(queries, &keys) / (M as f32);

        // Softmax on last dimension
        let token_weights = softmax(token_weights);

        // Get new tokens
        matmul(token_weights, &values)
    }
}
```
The difference is that the sequence length and even embedding size can be entirely different for tokens coming from the encoder, but attention allows us to handle this the same.

**Comment by coreylowman** [comment]
Nice work! I posted a concat_last in #43 that could be used here.

Also check out this commit https://github.com/coreylowman/dfdx/commit/480620fab3aaf29c964dc0a393b366787d70bd40, that's my last work on multi head attention. I was thinking we could do MHA with only reshapes basically.

**Comment by jafioti** [comment]
@coreylowman Actually I think I read the code wrong, it seems you aren't doing attention with other tokens. Your MultiHeadAttention only accepts one token at a time. What would it be attending to?

**Comment by jafioti** [comment]
This article explains it well: https://towardsdatascience.com/transformers-explained-visually-part-3-multi-head-attention-deep-dive-1c1ff1024853

**Comment by coreylowman** [comment]
Ahh yeah I see what you're saying. I missed the two different sequence lengths in your impl first review!

You can use linear per head by having the output of the linear be scaled by NUM_HEADS, and then reshape afterwards to `(..., NUM_HEADS, ...>` for multi head attention as far as I understood. I think that could be combined with your SHA if I'm understanding it correctly?

**Comment by coreylowman** [comment]
I didn't think that the heads operated on different slices of the inputs, from the paper it seemed like they all operated on the entire sequence? If that's the case I think we just need to have Q, K, V matrices output NUM_HEADS * <K or V> values, reshape into (..., NUM_HEADS, <K or V>), then do the attention stuff, then reshape into `NUM_HEADS * V`, and finally the W_OUT takes `NUM_HEADS * V`. Is that your interpretation as well?

I was referencing the timm vision transformer (https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py#L186), so idk if its different in NLP.

**Comment by jafioti** [comment]
My brain hurts figuring this out rn lol. I had an implementation in my old framework: https://github.com/Sidekick-AI/condor/blob/master/src/modules/transformer/base.rs that I know worked. I think you're right about using only one linear for all the heads, and most implementations even combine them into one linear layer for all query, key and value of size M * 3. I'm going to work on a solid implementation later. 

Also we need a transpose operator. I started working on one using a new `Transpose` trait, implemented for each possible pair of dims for each tensor, but I haven't worked out the most efficient way of transferring data yet.

**Comment by jafioti** [comment]
@coreylowman I think the main issue we face is doing this: 
```rust
let keys: Tensor2D<S, K>  = Tensor2D::zeros(); // This is just for the example
let keys: Tensor3D<S, {K / H}, H> = reshape(keys);
```
I'm not sure how to reshape them so that there is an added head dimension, so that the dot products can be done over each head and not mix.

**Comment by coreylowman** [comment]
@jafioti Wait what's the problem with that code snippet? I'm guessing the `{K / H}`, but I assumed that would be possible with `generic_const_exprs`?

**Comment by jafioti** [comment]
@coreylowman I get an error:
unconstrained generic constant
try adding a `where` bound using this expression: `where [(); { R::Array::NUM_ELEMENTS }]:`rustc
reshape.rs(18, 41): required by a bound in `tensor_ops::reshape::reshape`

I actually get a lot of these and I don't understand why. It wants a where clause that doesn't constrain an array to anything.

**Comment by coreylowman** [comment]
Ahh yeah that is generic_const_exprs for you 😄 that is by design as far as I understand. I always added a `: Sized`, but can't remember if that was necessary.

I don't really understand why either - it seems liek the compiler should just implicitly assume all of those? I dont' see a reason where you would want something that ISN'T valid.

**Comment by jafioti** [comment]
@coreylowman I got really stuck here and eventually figured out that ConstEq only works for some equality, and it doesn't evaluate expressions. For instance, this doesn't work:
```rust
impl<const A: usize, const B: usize, const C: usize, const D: usize, const E: usize>
    Reshape<Tensor3D<C, D, E>> for Tensor2D<A, B>
where
    ConstEq<{ A * B }, { C * D * E }>: ConstTrue,
{
    fn reshape(self) -> Tensor3D<C, D, E> {
        ...
    }
}
```
because it always sees `{ A * B }` as a different type than `{ C * D * E }`. Do you know a way around this?

**Comment by jafioti** [comment]
Yeah I can't seem to get this reshape to work at all. Can anyone else take a stab at it?
```rust
let keys: Tensor2D<S, K>  = Tensor2D::zeros(); // This is just for the example
let keys: Tensor3D<S, {K / H}, H> = reshape(keys);
```

**Comment by jafioti** [comment]
Alright after a lot of messing around, I finally got something that at least compiles:  https://github.com/coreylowman/dfdx/commit/733b9c6190880c8b260508ce1d18b1a06cce34a1

The trick is I defined a `Reshape` trait and implemented it for every dimension pair (Tensor1D to Tensor2D, Tensor2D to Tensor1D, etc.). This allowed each implementation to have it's own asserts to check whether the products of the dims where equal, to ensure the total number of elements are equal.

I then used this new reshape to implement `MultiHeadAttention`, but currently it only uses reshape and I'm fairly certain it has to use transpose to keep the numbers in the right spots, so I'll be working on that next.

A step in the right direction, but not there yet.

**Comment by vikigenius** [comment]
That seems awfully inconvenient if we have to implement it for every pair. Is there no way to have it be generic over tensors like the original implementation.

**Comment by jafioti** [comment]
It wasn't working for me, was throwing the weird errors you were seeing. Sure that would be better, but if you look at mine, I don't manually implement it, I just use a macro.

**Comment by jafioti** [comment]
Yeah I for sure need transpose here because my outputs are different than pytorch's with the same weights.

**Comment by jafioti** [comment]
I finally got a working implementation of a Transformer Encoder! 
https://github.com/jafioti/dfdx/commit/cbec5c4f7e65151f868203033a487f5a8e8c0193

I tested it against minGPT and got correct activations. Currently working on batched forward impls, then I think it might be ready to merge.

**Comment by coreylowman** [comment]
@jafioti WOW awesome work! 🎉 

Do you have a better sense of what the issue is with reshape? I see you have the asserts both ways (Assert<A, B> & Assert<B, A>), and then also had to had multiply all the elements instead of using NUM_ELEMENTS.

**Comment by jafioti** [comment]
@coreylowman Basically I needed a where clause in the forward function of self attention to specify that the num elements were the same, but when I put that in, I got a "cycle detected" error which I don't know why. Ended up with this impl, but I actually don't think it's too bad because the impls are done with a macro. Of course a single reshape function would be better, but for now it works.

I also had issues with ConstEq, so I switched to Assert<>, but there's a separate gh issue for that.
